diff -Naur ffmpeg-7.1.4.old/libavformat/avformat.h ffmpeg-7.1.4/libavformat/avformat.h
--- ffmpeg-7.1.4.old/libavformat/avformat.h	2026-05-05 08:18:37.492469214 +0200
+++ ffmpeg-7.1.4/libavformat/avformat.h	2026-05-05 08:18:38.189510338 +0200
@@ -1202,6 +1202,10 @@
 
 struct AVCodecParserContext *av_stream_get_parser(const AVStream *s);
 
+// Chromium: We use the internal field first_dts vvv
+int64_t    av_stream_get_first_dts(const AVStream *st);
+// Chromium: We use the internal field first_dts ^^^
+
 #define AV_PROGRAM_RUNNING 1
 
 /**
diff -Naur ffmpeg-7.1.4.old/libavformat/utils.c ffmpeg-7.1.4/libavformat/utils.c
--- ffmpeg-7.1.4.old/libavformat/utils.c	2026-05-05 08:18:37.495469252 +0200
+++ ffmpeg-7.1.4/libavformat/utils.c	2026-05-05 08:18:38.190509381 +0200
@@ -44,6 +44,13 @@
  * various utility functions for use within FFmpeg
  */
 
+// Chromium: We use the internal field first_dts vvv
+int64_t av_stream_get_first_dts(const AVStream *st)
+{
+  return cffstream(st)->first_dts;
+}
+// Chromium: We use the internal field first_dts ^^^
+
 /* an arbitrarily chosen "sane" max packet size -- 50M */
 #define SANE_CHUNK_SIZE (50000000)
 
