diff -Naur ffmpeg-7.1.2.old/libavformat/avformat.h ffmpeg-7.1.2/libavformat/avformat.h
--- ffmpeg-7.1.2.old/libavformat/avformat.h	2025-09-16 15:39:29.189893029 +0200
+++ ffmpeg-7.1.2/libavformat/avformat.h	2025-09-16 15:39:30.139792825 +0200
@@ -1202,6 +1202,10 @@
 
 struct AVCodecParserContext *av_stream_get_parser(const AVStream *s);
 
+// Chromium: We use the internal field first_dts vvv
+int64_t    av_stream_get_first_dts(const AVStream *st);
+// Chromium: We use the internal field first_dts ^^^
+
 #define AV_PROGRAM_RUNNING 1
 
 /**
diff -Naur ffmpeg-7.1.2.old/libavformat/utils.c ffmpeg-7.1.2/libavformat/utils.c
--- ffmpeg-7.1.2.old/libavformat/utils.c	2025-09-16 15:39:29.193893061 +0200
+++ ffmpeg-7.1.2/libavformat/utils.c	2025-09-16 15:39:30.140432771 +0200
@@ -44,6 +44,13 @@
  * various utility functions for use within FFmpeg
  */
 
+// Chromium: We use the internal field first_dts vvv
+int64_t av_stream_get_first_dts(const AVStream *st)
+{
+  return cffstream(st)->first_dts;
+}
+// Chromium: We use the internal field first_dts ^^^
+
 /* an arbitrarily chosen "sane" max packet size -- 50M */
 #define SANE_CHUNK_SIZE (50000000)
 
