diff -Naur ffmpeg-9.0.1.old/configure ffmpeg-9.0.1/configure
--- ffmpeg-9.0.1.old/configure	2026-09-18 07:59:39.300503532 +0200
+++ ffmpeg-9.0.1/configure	2026-09-18 07:59:39.465505065 +0200
@@ -5057,7 +5057,7 @@
 
 if enabled cuda_nvcc; then
     if $nvcc $nvccflags_default 2>&1 | grep -qi unsupported; then
-        nvccflags_default="-gencode arch=compute_60,code=sm_60 -O2"
+        nvccflags_default="-gencode arch=compute_75,code=sm_75 -O2"
     fi
     if $nvcc $nvccflags_default 2>&1 | grep -qi unsupported; then
         nvccflags_default="-gencode arch=compute_75,code=sm_75 -O2"
@@ -7196,7 +7196,7 @@
     nvccflags=$nvccflags_default
 fi
 
-nvccflags="$nvccflags -std=c++11"
+nvccflags="$nvccflags -std=c++14"
 
 if enabled x86_64 || enabled ppc64 || enabled aarch64; then
     nvccflags="$nvccflags -m64"
