diff -Naur ffmpeg-8.0.1.old/libavformat/avformat.h ffmpeg-8.0.1/libavformat/avformat.h
--- ffmpeg-8.0.1.old/libavformat/avformat.h	2026-02-12 18:12:37.721511421 +0100
+++ ffmpeg-8.0.1/libavformat/avformat.h	2026-02-12 18:12:38.554767892 +0100
@@ -1179,6 +1179,10 @@
 
 struct AVCodecParserContext *av_stream_get_parser(const AVStream *s);
 
+// Chromium: We use the internal field first_dts vvv
+int64_t    av_stream_get_first_dts(const AVStream *st);
+// Chromium: We use the internal field first_dts ^^^
+
 #define AV_PROGRAM_RUNNING 1
 
 /**
diff -Naur ffmpeg-8.0.1.old/libavformat/utils.c ffmpeg-8.0.1/libavformat/utils.c
--- ffmpeg-8.0.1.old/libavformat/utils.c	2026-02-12 18:12:37.725511469 +0100
+++ ffmpeg-8.0.1/libavformat/utils.c	2026-02-12 18:12:38.555414540 +0100
@@ -47,6 +47,13 @@
  * various utility functions for use within FFmpeg
  */
 
+// Chromium: We use the internal field first_dts vvv
+int64_t av_stream_get_first_dts(const AVStream *st)
+{
+  return cffstream(st)->first_dts;
+}
+// Chromium: We use the internal field first_dts ^^^
+
 /* an arbitrarily chosen "sane" max packet size -- 50M */
 #define SANE_CHUNK_SIZE (50000000)
 
