diff -Naur ffmpeg-9.0.1.old/libavcodec/amfenc.c ffmpeg-9.0.1/libavcodec/amfenc.c
--- ffmpeg-9.0.1.old/libavcodec/amfenc.c	2026-09-18 07:59:38.256493829 +0200
+++ ffmpeg-9.0.1/libavcodec/amfenc.c	2026-09-18 07:59:39.665414509 +0200
@@ -119,10 +119,6 @@
     else
         pix_fmt = avctx->pix_fmt;
 
-    if (pix_fmt == AV_PIX_FMT_P010) {
-        AMF_RETURN_IF_FALSE(ctx, amf_device_ctx->version >= AMF_MAKE_FULL_VERSION(1, 4, 32, 0), AVERROR_UNKNOWN, "10-bit encoder is not supported by AMD GPU drivers versions lower than 23.30.\n");
-    }
-
     ctx->format = av_av_to_amf_format(pix_fmt);
     AMF_RETURN_IF_FALSE(ctx, ctx->format != AMF_SURFACE_UNKNOWN, AVERROR(EINVAL),
                         "Format %s is not supported\n", av_get_pix_fmt_name(pix_fmt));
diff -Naur ffmpeg-9.0.1.old/libavcodec/ccaption_dec.c ffmpeg-9.0.1/libavcodec/ccaption_dec.c
--- ffmpeg-9.0.1.old/libavcodec/ccaption_dec.c	2026-09-18 07:59:38.313494359 +0200
+++ ffmpeg-9.0.1/libavcodec/ccaption_dec.c	2026-09-18 07:59:39.652535529 +0200
@@ -893,12 +893,13 @@
 
         if (ctx->buffer[bidx].str[0] || ctx->real_time) {
             ff_dlog(avctx, "cdp writing data (%s)\n", ctx->buffer[bidx].str);
-            start_time = ctx->buffer_time[0];
-            sub->pts = start_time;
-            end_time = ctx->buffer_time[1];
-            if (!ctx->real_time)
+            if (!ctx->real_time) {
+                start_time = ctx->buffer_time[0];
+                sub->pts = start_time;
+                end_time = ctx->buffer_time[1];
                 sub->end_display_time = av_rescale_q(end_time - start_time,
                                                      AV_TIME_BASE_Q, ms_tb);
+            }
             else
                 sub->end_display_time = -1;
             ret = ff_ass_add_rect2(sub, ctx->buffer[bidx].str, ctx->readorder++, 0, NULL, NULL, &nb_rect_allocated);
diff -Naur ffmpeg-9.0.1.old/libavcodec/dvdsubdec.c ffmpeg-9.0.1/libavcodec/dvdsubdec.c
--- ffmpeg-9.0.1.old/libavcodec/dvdsubdec.c	2026-09-18 07:59:38.227493560 +0200
+++ ffmpeg-9.0.1/libavcodec/dvdsubdec.c	2026-09-18 07:59:39.651622848 +0200
@@ -45,6 +45,8 @@
   int      buf_size;
   int      forced_subs_only;
   uint8_t  used_color[256];
+  int64_t  pts;
+  int      output_empty_rects;
 } DVDSubContext;
 
 static void yuv_a_to_rgba(const uint8_t *ycbcr, const uint8_t *alpha, uint32_t *rgba, int num_values)
@@ -230,7 +232,10 @@
     uint32_t size;
     int64_t offset1, offset2;
 
-    if (buf_size < 10)
+    if (buf_size < 2)
+        return AVERROR(EAGAIN);
+
+    if (buf_size == 2 && AV_RB16(buf) == 0)
         return -1;
 
     if (AV_RB16(buf) == 0) {   /* HD subpicture with 4-byte offsets */
@@ -243,15 +248,21 @@
         cmd_pos = 2;
     }
 
+    if (big_offsets && buf_size < 6)
+        return AVERROR(EAGAIN);
+
     size = READ_OFFSET(buf + (big_offsets ? 2 : 0));
-    cmd_pos = READ_OFFSET(buf + cmd_pos);
+    if (size == 0)
+        return -1;
 
-    if (cmd_pos < 0 || cmd_pos > buf_size - 2 - offset_size) {
-        if (cmd_pos > size) {
-            av_log(logctx, AV_LOG_ERROR, "Discarding invalid packet\n");
-            return 0;
-        }
+    if (buf_size < size)
         return AVERROR(EAGAIN);
+
+    cmd_pos = READ_OFFSET(buf + cmd_pos);
+
+    if (cmd_pos < 0 || cmd_pos > size) {
+        av_log(ctx, AV_LOG_ERROR, "Discarding invalid packet\n");
+        return AVERROR_INVALIDDATA;
     }
 
     while (cmd_pos > 0 && cmd_pos < buf_size - 2 - offset_size) {
@@ -524,10 +535,13 @@
     int appended = 0;
     int is_menu;
 
+    if (ctx->pts == AV_NOPTS_VALUE && sub->pts != AV_NOPTS_VALUE)
+        ctx->pts = sub->pts;
     if (ctx->buf_size) {
         int ret = append_to_cached_buf(avctx, buf, buf_size);
         if (ret < 0) {
             *data_size = 0;
+            ctx->pts = AV_NOPTS_VALUE;
             return ret;
         }
         buf = ctx->buf;
@@ -538,7 +552,12 @@
     is_menu = decode_dvd_subtitles(avctx, ctx, sub, buf, buf_size);
     if (is_menu == AVERROR(EAGAIN)) {
         *data_size = 0;
-        return appended ? 0 : append_to_cached_buf(avctx, buf, buf_size);
+        int ret = appended ? 0 : append_to_cached_buf(avctx, buf, buf_size);
+        if (ret < 0) {
+            ctx->pts = AV_NOPTS_VALUE;
+            return ret;
+        }
+        return buf_size;
     }
 
     if (is_menu < 0) {
@@ -547,9 +566,10 @@
         reset_rects(sub);
         *data_size = 0;
 
+        ctx->pts = AV_NOPTS_VALUE;
         return buf_size;
     }
-    if (!is_menu && find_smallest_bounding_rectangle(ctx, sub) == 0)
+    if (!is_menu && !ctx->output_empty_rects && find_smallest_bounding_rectangle(ctx, sub) == 0)
         goto no_subtitle;
 
     if (ctx->forced_subs_only && !(sub->rects[0]->flags & AV_SUBTITLE_FLAG_FORCED))
@@ -557,6 +577,8 @@
 
     ctx->buf_size = 0;
     *data_size = 1;
+    sub->pts = ctx->pts;
+    ctx->pts = AV_NOPTS_VALUE;
     return buf_size;
 }
 
@@ -683,6 +705,7 @@
             av_log(avctx, AV_LOG_DEBUG, " 0x%06"PRIx32, ctx->palette[i]);
         av_log(avctx, AV_LOG_DEBUG, "\n");
     }
+    ctx->pts = AV_NOPTS_VALUE;
 
     return 0;
 }
@@ -699,6 +722,7 @@
     { "palette", "set the global palette", OFFSET(palette_str), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, SD },
     { "ifo_palette", "obtain the global palette from .IFO file", OFFSET(ifo_str), AV_OPT_TYPE_STRING, { .str = NULL }, 0, 0, SD },
     { "forced_subs_only", "Only show forced subtitles", OFFSET(forced_subs_only), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, SD},
+    { "output_empty_rects", "Output subtitles with empty or fully transparent rects", OFFSET(output_empty_rects), AV_OPT_TYPE_BOOL, {.i64 = 0}, 0, 1, SD},
     { NULL }
 };
 static const AVClass dvdsub_class = {
diff -Naur ffmpeg-9.0.1.old/libavcodec/itut35.c ffmpeg-9.0.1/libavcodec/itut35.c
--- ffmpeg-9.0.1.old/libavcodec/itut35.c	2026-09-18 07:59:38.310494331 +0200
+++ ffmpeg-9.0.1/libavcodec/itut35.c	2026-09-18 07:59:39.655201876 +0200
@@ -355,6 +355,9 @@
     }
 
     if (metadata.dovi) {
+        AVBufferRef *rpu_buf;
+        AVFrameSideData *rpu;
+
         AVFrameSideData *sd = av_frame_new_side_data_from_buf(frame, AV_FRAME_DATA_DOVI_METADATA,
                                                               metadata.dovi);
         if (!sd) {
@@ -362,6 +365,22 @@
             return AVERROR(ENOMEM);
         }
         metadata.dovi = NULL;
+
+        rpu_buf = av_buffer_alloc(itut_t35->payload_size + 6);
+        if (rpu_buf) {
+            PutByteContext p;
+            bytestream2_init_writer(&p, rpu_buf->data, 6);
+            bytestream2_put_be16(&p, itut_t35->provider_code);
+            bytestream2_put_be32(&p, itut_t35->provider_oriented_code);
+            memcpy(rpu_buf->data + 6, itut_t35->payload, itut_t35->payload_size);
+            rpu = av_frame_new_side_data_from_buf(frame, AV_FRAME_DATA_DOVI_RPU_BUFFER_T35, rpu_buf);
+            if (!rpu) {
+                av_buffer_unref(&rpu_buf);
+                return AVERROR(ENOMEM);
+            }
+        } else {
+            return AVERROR(ENOMEM);
+        }
     }
 
     if (metadata.hdr_smpte2094_app5) {
diff -Naur ffmpeg-9.0.1.old/libavcodec/qsvenc_av1.c ffmpeg-9.0.1/libavcodec/qsvenc_av1.c
--- ffmpeg-9.0.1.old/libavcodec/qsvenc_av1.c	2026-09-18 07:59:38.297494210 +0200
+++ ffmpeg-9.0.1/libavcodec/qsvenc_av1.c	2026-09-18 07:59:39.658205769 +0200
@@ -189,6 +189,10 @@
     { "tile_cols",  "Number of columns for tiled encoding",   OFFSET(qsv.tile_cols),    AV_OPT_TYPE_INT, { .i64 = 0 }, 0, UINT16_MAX, VE },
     { "tile_rows",  "Number of rows for tiled encoding",      OFFSET(qsv.tile_rows),    AV_OPT_TYPE_INT, { .i64 = 0 }, 0, UINT16_MAX, VE },
     { "look_ahead_depth", "Depth of look ahead in number frames, available when extbrc option is enabled", OFFSET(qsv.look_ahead_depth), AV_OPT_TYPE_INT, { .i64 = 0 }, 0, 100, VE },
+#if QSV_HAVE_EXT_AV1_SCC
+    { "palette_mode", "Enable palette mode of Screen Content Tool for encoding", OFFSET(qsv.palette_mode), AV_OPT_TYPE_BOOL, { .i64 = 0}, 0, 1, VE},
+    { "intrabc", "Enable intra block copy of Screen Content Tool for encoding", OFFSET(qsv.intrabc), AV_OPT_TYPE_BOOL, { .i64 = 0}, 0, 1, VE},
+#endif
     { NULL },
 };
 
diff -Naur ffmpeg-9.0.1.old/libavcodec/qsvenc.c ffmpeg-9.0.1/libavcodec/qsvenc.c
--- ffmpeg-9.0.1.old/libavcodec/qsvenc.c	2026-09-18 07:59:38.313494359 +0200
+++ ffmpeg-9.0.1/libavcodec/qsvenc.c	2026-09-18 07:59:39.657937744 +0200
@@ -492,6 +492,9 @@
     mfxExtAV1BitstreamParam *av1_bs_param = (mfxExtAV1BitstreamParam *)coding_opts[1];
     mfxExtCodingOption2 *co2 = (mfxExtCodingOption2*)coding_opts[2];
     mfxExtCodingOption3 *co3 = (mfxExtCodingOption3*)coding_opts[3];
+#if QSV_HAVE_EXT_AV1_SCC
+    mfxExtAV1ScreenContentTools *scc = (mfxExtAV1ScreenContentTools*)coding_opts[4];
+#endif
 
     av_log(avctx, AV_LOG_VERBOSE, "profile: %s; level: %"PRIu16"\n",
            print_profile(avctx->codec_id, info->CodecProfile), info->CodecLevel);
@@ -564,6 +567,13 @@
            print_threestate(av1_bs_param->WriteIVFHeaders));
     av_log(avctx, AV_LOG_VERBOSE, "LowDelayBRC: %s\n", print_threestate(co3->LowDelayBRC));
     av_log(avctx, AV_LOG_VERBOSE, "MaxFrameSize: %d;\n", co2->MaxFrameSize);
+#if QSV_HAVE_EXT_AV1_SCC
+    if (scc) {
+        av_log(avctx, AV_LOG_VERBOSE,
+               "Palette: %s; IntraBlockCopy: %s\n",
+               print_threestate(scc->Palette), print_threestate(scc->IntraBlockCopy));
+    }
+#endif
 }
 #endif
 
@@ -1284,6 +1294,28 @@
     }
 #endif
 
+#if QSV_HAVE_EXT_AV1_SCC
+    if (q->palette_mode || q->intrabc) {
+        if (QSV_RUNTIME_VERSION_ATLEAST(q->ver, 2, 13)) {
+            if (q->param.mfx.CodecId != MFX_CODEC_AV1) {
+                av_log(avctx, AV_LOG_ERROR, "Not supported encoder for Screen Content Tool Encode. "
+                                            "Supported: av1_qsv \n");
+                return AVERROR_UNKNOWN;
+            }
+
+            q->extsccparam.Header.BufferId = MFX_EXTBUFF_AV1_SCREEN_CONTENT_TOOLS;
+            q->extsccparam.Header.BufferSz = sizeof(q->extsccparam);
+            q->extsccparam.Palette = q->palette_mode ? MFX_CODINGOPTION_ON : MFX_CODINGOPTION_OFF;
+            q->extsccparam.IntraBlockCopy = q->intrabc ? MFX_CODINGOPTION_ON : MFX_CODINGOPTION_OFF;
+            q->extparam_internal[q->nb_extparam_internal++] = (mfxExtBuffer *)&q->extsccparam;
+        } else {
+            av_log(avctx, AV_LOG_ERROR,
+                   "This version of runtime doesn't support Screen Content Tool Encode\n");
+            return AVERROR_UNKNOWN;
+        }
+    }
+#endif
+
     if (!check_enc_param(avctx,q)) {
         av_log(avctx, AV_LOG_ERROR,
                "some encoding parameters are not supported by the QSV "
@@ -1391,11 +1423,21 @@
         .Header.BufferSz = sizeof(co3),
     };
 
+#if QSV_HAVE_EXT_AV1_SCC
+    mfxExtAV1ScreenContentTools scc_buf = {
+        .Header.BufferId = MFX_EXTBUFF_AV1_SCREEN_CONTENT_TOOLS,
+        .Header.BufferSz = sizeof(scc_buf),
+    };
+#endif
+
     mfxExtBuffer *ext_buffers[] = {
         (mfxExtBuffer*)&av1_extend_tile_buf,
         (mfxExtBuffer*)&av1_bs_param,
         (mfxExtBuffer*)&co2,
         (mfxExtBuffer*)&co3,
+#if QSV_HAVE_EXT_AV1_SCC
+        (mfxExtBuffer*)&scc_buf,
+#endif
     };
 
     if (!QSV_RUNTIME_VERSION_ATLEAST(q->ver, 2, 5)) {
@@ -1844,6 +1886,13 @@
         return ret;
     }
 
+    // Update AVCodecContext with actual encoding parameters
+    mfxInfoMFX *info = &q->param.mfx;
+    avctx->has_b_frames = 0;
+    if (info->GopRefDist > 1) {
+        avctx->has_b_frames = info->GopRefDist - 1;
+    }
+
     q->avctx = avctx;
 
     return 0;
diff -Naur ffmpeg-9.0.1.old/libavcodec/qsvenc.h ffmpeg-9.0.1/libavcodec/qsvenc.h
--- ffmpeg-9.0.1.old/libavcodec/qsvenc.h	2026-09-18 07:59:38.221493504 +0200
+++ ffmpeg-9.0.1/libavcodec/qsvenc.h	2026-09-18 07:59:39.658128273 +0200
@@ -38,6 +38,7 @@
 
 #define QSV_HAVE_EXT_VP9_TILES QSV_VERSION_ATLEAST(1, 29)
 #define QSV_HAVE_EXT_AV1_PARAM QSV_VERSION_ATLEAST(2, 5)
+#define QSV_HAVE_EXT_AV1_SCC   QSV_VERSION_ATLEAST(2, 13)
 
 #if defined(_WIN32) || defined(__CYGWIN__)
 #define QSV_HAVE_AVBR   1
@@ -188,10 +189,13 @@
     mfxFrameSurface1       **opaque_surfaces;
     AVBufferRef             *opaque_alloc_buf;
 #endif
+#if QSV_HAVE_EXT_AV1_SCC
+    mfxExtAV1ScreenContentTools extsccparam;
+#endif
 
     mfxExtVideoSignalInfo extvsi;
 
-    mfxExtBuffer  *extparam_internal[5 + (QSV_HAVE_MF * 2) + (QSV_HAVE_EXT_AV1_PARAM * 2) + QSV_HAVE_HE];
+    mfxExtBuffer  *extparam_internal[5 + (QSV_HAVE_MF * 2) + (QSV_HAVE_EXT_AV1_PARAM * 2) + QSV_HAVE_HE + QSV_HAVE_EXT_AV1_SCC];
     int         nb_extparam_internal;
 
     mfxExtBuffer  **extparam_str;
@@ -319,6 +323,8 @@
     int dual_gfx;
 
     AVDictionary *qsv_params;
+    int palette_mode;
+    int intrabc;
 } QSVEncContext;
 
 int ff_qsv_enc_init(AVCodecContext *avctx, QSVEncContext *q);
diff -Naur ffmpeg-9.0.1.old/libavcodec/videotoolbox.c ffmpeg-9.0.1/libavcodec/videotoolbox.c
--- ffmpeg-9.0.1.old/libavcodec/videotoolbox.c	2026-09-18 07:59:38.253493801 +0200
+++ ffmpeg-9.0.1/libavcodec/videotoolbox.c	2026-09-18 07:59:39.664225187 +0200
@@ -120,7 +120,6 @@
 
 static int videotoolbox_postproc_frame(void *avctx, AVFrame *frame)
 {
-    int ret;
     VTHWFrame *ref = (VTHWFrame *)frame->buf[0]->data;
 
     if (!ref->pixbuf) {
@@ -134,9 +133,6 @@
     frame->crop_top = 0;
     frame->crop_bottom = 0;
 
-    if ((ret = av_vt_pixbuf_set_attachments(avctx, ref->pixbuf, frame)) < 0)
-        return ret;
-
     frame->data[3] = (uint8_t*)ref->pixbuf;
 
     if (ref->hw_frames_ctx) {
@@ -822,7 +818,7 @@
 #if TARGET_OS_IPHONE
     CFDictionarySetValue(buffer_attributes, kCVPixelBufferOpenGLESCompatibilityKey, kCFBooleanTrue);
 #else
-    CFDictionarySetValue(buffer_attributes, kCVPixelBufferIOSurfaceOpenGLTextureCompatibilityKey, kCFBooleanTrue);
+    CFDictionarySetValue(buffer_attributes, kCVPixelBufferMetalCompatibilityKey, kCFBooleanTrue);
 #endif
 
     CFRelease(io_surface_properties);
@@ -964,6 +960,23 @@
         break;
     }
 
+#if ARCH_X86_64
+    if (avctx->codec_id == AV_CODEC_ID_H264 &&
+        avctx->sw_pix_fmt == AV_PIX_FMT_YUV420P10)
+    {
+        // 10-bit H.264 is not supported on x86_64
+        return AVERROR(ENOSYS);
+    }
+#endif
+
+    if (avctx->codec_id == AV_CODEC_ID_H264 &&
+        (avctx->level == 61 || avctx->level == 62))
+    {
+        // H.264 Level 6.1 and 6.2 can't be
+        // decoded properly
+        return AVERROR(ENOSYS);
+    }
+
 #if defined(MAC_OS_X_VERSION_10_9) && !TARGET_OS_IPHONE && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_9) && AV_HAS_BUILTIN(__builtin_available)
     if (avctx->codec_id == AV_CODEC_ID_PRORES || avctx->codec_id == AV_CODEC_ID_PRORES_RAW) {
         if (__builtin_available(macOS 10.9, *)) {
diff -Naur ffmpeg-9.0.1.old/libavfilter/vf_vpp_qsv.c ffmpeg-9.0.1/libavfilter/vf_vpp_qsv.c
--- ffmpeg-9.0.1.old/libavfilter/vf_vpp_qsv.c	2026-09-18 07:59:38.420495353 +0200
+++ ffmpeg-9.0.1/libavfilter/vf_vpp_qsv.c	2026-09-18 07:59:39.659577435 +0200
@@ -494,9 +494,9 @@
     outvsi_conf.Header.BufferId          = MFX_EXTBUFF_VIDEO_SIGNAL_INFO_OUT;
     outvsi_conf.Header.BufferSz          = sizeof(mfxExtVideoSignalInfo);
     outvsi_conf.VideoFullRange           = (out->color_range == AVCOL_RANGE_JPEG);
-    outvsi_conf.ColourPrimaries          = (out->color_primaries == AVCOL_PRI_UNSPECIFIED) ? AVCOL_PRI_BT709 : out->color_primaries;
-    outvsi_conf.TransferCharacteristics  = (out->color_trc == AVCOL_TRC_UNSPECIFIED) ? AVCOL_TRC_BT709 : out->color_trc;
-    outvsi_conf.MatrixCoefficients       = (out->colorspace == AVCOL_SPC_UNSPECIFIED) ? AVCOL_SPC_BT709 : out->colorspace;
+    outvsi_conf.ColourPrimaries          = (out->color_primaries == AVCOL_PRI_UNSPECIFIED) ? invsi_conf.ColourPrimaries : out->color_primaries;
+    outvsi_conf.TransferCharacteristics  = (out->color_trc == AVCOL_TRC_UNSPECIFIED) ? invsi_conf.TransferCharacteristics : out->color_trc;
+    outvsi_conf.MatrixCoefficients       = (out->colorspace == AVCOL_SPC_UNSPECIFIED) ? invsi_conf.MatrixCoefficients : out->colorspace;
     outvsi_conf.ColourDescriptionPresent = 1;
 
     if (memcmp(&vpp->invsi_conf, &invsi_conf, sizeof(mfxExtVideoSignalInfo)) ||
diff -Naur ffmpeg-9.0.1.old/libavformat/isom.h ffmpeg-9.0.1/libavformat/isom.h
--- ffmpeg-9.0.1.old/libavformat/isom.h	2026-09-18 07:59:38.470495818 +0200
+++ ffmpeg-9.0.1/libavformat/isom.h	2026-09-18 07:59:39.640937844 +0200
@@ -286,6 +286,9 @@
         MOVEncryptionIndex *encryption_index;
     } cenc;
 
+    int has_fallback;   // Audio fallback track
+    int fallback;
+
     struct IAMFDemuxContext *iamf;
     int iamf_stream_offset;
 } MOVStreamContext;
diff -Naur ffmpeg-9.0.1.old/libavformat/matroskaenc.c ffmpeg-9.0.1/libavformat/matroskaenc.c
--- ffmpeg-9.0.1.old/libavformat/matroskaenc.c	2026-09-18 07:59:38.484495948 +0200
+++ ffmpeg-9.0.1/libavformat/matroskaenc.c	2026-09-18 07:59:39.653506812 +0200
@@ -3066,6 +3066,16 @@
     case AV_CODEC_ID_AAC:
         if (side_data_size && mkv->track.bc) {
             int output_sample_rate = 0;
+            if (par->extradata && par->extradata_size) {
+                if (par->extradata_size != side_data_size ||
+                    memcmp(par->extradata, side_data, side_data_size)) {
+                    av_log(s, AV_LOG_ERROR, "Error, AAC extradata changed mid-stream.\n");
+                    return AVERROR(EINVAL);
+                } else {
+                    // Already written
+                    break;
+                }
+            }
             ret = get_aac_sample_rates(s, mkv, side_data, side_data_size,
                                        &track->sample_rate, &output_sample_rate);
             if (ret < 0)
diff -Naur ffmpeg-9.0.1.old/libavformat/mov.c ffmpeg-9.0.1/libavformat/mov.c
--- ffmpeg-9.0.1.old/libavformat/mov.c	2026-09-18 07:59:38.464495762 +0200
+++ ffmpeg-9.0.1/libavformat/mov.c	2026-09-18 07:59:39.644506729 +0200
@@ -58,6 +58,7 @@
 #include "libavcodec/mpegaudiodecheader.h"
 #include "libavcodec/mlp_parse.h"
 #include "avformat.h"
+#include "avlanguage.h"
 #include "internal.h"
 #include "avio_internal.h"
 #include "demux.h"
@@ -132,6 +133,33 @@
     return 0;
 }
 
+static int mov_metadata_int16_no_padding(MOVContext *c, AVIOContext *pb,
+                                         unsigned len, const char *key)
+{
+    c->fc->event_flags |= AVFMT_EVENT_FLAG_METADATA_UPDATED;
+    av_dict_set_int(&c->fc->metadata, key, avio_rb16(pb), 0);
+
+    return 0;
+}
+
+static int mov_metadata_int32_no_padding(MOVContext *c, AVIOContext *pb,
+                                         unsigned len, const char *key)
+{
+    c->fc->event_flags |= AVFMT_EVENT_FLAG_METADATA_UPDATED;
+    av_dict_set_int(&c->fc->metadata, key, avio_rb32(pb), 0);
+
+    return 0;
+}
+
+static int mov_metadata_int64_no_padding(MOVContext *c, AVIOContext *pb,
+                                         unsigned len, const char *key)
+{
+    c->fc->event_flags |= AVFMT_EVENT_FLAG_METADATA_UPDATED;
+    av_dict_set_int(&c->fc->metadata, key, avio_rb64(pb), 0);
+
+    return 0;
+}
+
 static int mov_metadata_gnre(MOVContext *c, AVIOContext *pb,
                              unsigned len, const char *key)
 {
@@ -343,6 +371,73 @@
     return 0;
 }
 
+static int mov_read_3gp_udta_tag(MOVContext *c, AVIOContext *pb, MOVAtom atom)
+{
+    const char *key;
+    AVDictionary *metadata;
+    uint16_t langcode = 0;
+    char key2[32], language[4] = {0};
+    uint32_t str_size, version;
+    char *str;
+    if (atom.size < 6)
+        return AVERROR_INVALIDDATA;
+    switch (atom.type) {
+    case MKTAG( 'a','l','b','m'): key = "album"; break;
+    case MKTAG( 'a','u','t','h'): key = "author"; break;
+    case MKTAG( 'c','p','r','t'): key = "copyright"; break;
+    case MKTAG( 'd','s','c','p'): key = "comment"; break;
+    case MKTAG( 'g','n','r','e'): key = "genre"; break;
+    case MKTAG( 'p','e','r','f'): key = "artist"; break;
+    case MKTAG( 't','i','t','l'): key = "title"; break;
+    case MKTAG( 'y','r','r','c'): key = "date"; break;
+    default: return 0;
+    }
+    version = avio_rb32(pb); // version + flags
+    if (version != 0)
+        av_log(c->fc, AV_LOG_WARNING, "udta %s unknown version number: %u\n", key, version);
+    if (MKTAG( 'y','r','r','c') == atom.type) {
+        int year;
+        year = avio_rb16(pb);
+        str = av_asprintf("%d", year);
+        if (!str)
+            return AVERROR(ENOMEM);
+    } else {
+        int ret;
+        const char *tmp;
+        langcode = avio_rb16(pb);
+        ff_mov_lang_to_iso639(langcode, language);
+        tmp = ff_convert_lang_to(language, AV_LANG_ISO639_2_BIBL);
+        if (!tmp)
+           av_log(c->fc, AV_LOG_WARNING, "udta %s unknown language code: %u\n", key, langcode);
+        str_size = atom.size - 6;
+        if (str_size <= 0 || str_size >= INT_MAX/2)
+            return AVERROR_INVALIDDATA;
+        str = av_mallocz(str_size + 1);
+        if (!str)
+            return AVERROR(ENOMEM);
+        ret = ffio_read_size(pb, str, str_size);
+        if (ret < 0) {
+            av_free(str);
+            return ret;
+        }
+        str[str_size] = 0;
+    }
+    if (c->trak_index < 0) {
+        metadata = c->fc->metadata;
+        c->fc->event_flags |= AVFMT_EVENT_FLAG_METADATA_UPDATED;
+    }
+    else {
+        metadata = c->fc->streams[c->trak_index]->metadata;
+    }
+    av_dict_set(&metadata, key, str, 0);
+    if (*language && strcmp(language, "und")) {
+        snprintf(key2, sizeof(key2), "%s-%s", key, language);
+        av_dict_set(&metadata, key2, str, 0);
+    }
+    av_freep(&str);
+    return 0;
+}
+
 static int mov_read_udta_string(MOVContext *c, AVIOContext *pb, MOVAtom atom)
 {
     char tmp_key[AV_FOURCC_MAX_STRING_SIZE] = {0};
@@ -371,36 +466,71 @@
     case MKTAG( 'a','k','I','D'): key = "account_type";
         parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 'a','p','I','D'): key = "account_id"; break;
+    case MKTAG( 'a','t','I','D'): key = "artist_id";
+        parse = mov_metadata_int32_no_padding; break;
     case MKTAG( 'c','a','t','g'): key = "category"; break;
+    case MKTAG( 'c','m','I','D'): key = "composer_id";
+        parse = mov_metadata_int32_no_padding; break;
+    case MKTAG( 'c','n','I','D'): key = "content_id";
+        parse = mov_metadata_int32_no_padding; break;
     case MKTAG( 'c','p','i','l'): key = "compilation";
         parse = mov_metadata_int8_no_padding; break;
-    case MKTAG( 'c','p','r','t'): key = "copyright"; break;
+    case MKTAG( 'c','p','r','t'):
+        key = "copyright";
+        if (!c->itunes_metadata) {
+            int64_t pos = avio_tell(pb);
+            int ret = mov_read_3gp_udta_tag(c, pb, atom);
+            if (ret != AVERROR_INVALIDDATA)
+                return ret;
+            avio_seek(pb, pos, SEEK_SET);
+        }
+        break;
     case MKTAG( 'd','e','s','c'): key = "description"; break;
     case MKTAG( 'd','i','s','k'): key = "disc";
         parse = mov_metadata_track_or_disc_number; break;
     case MKTAG( 'e','g','i','d'): key = "episode_uid";
         parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 'F','I','R','M'): key = "firmware"; raw = 1; break;
-    case MKTAG( 'g','n','r','e'): key = "genre";
-        parse = mov_metadata_gnre; break;
+    case MKTAG( 'g','e','I','D'): key = "genre_id";
+        parse = mov_metadata_int32_no_padding; break;
+    case MKTAG( 'g','n','r','e'):
+        key = "genre";
+        parse = mov_metadata_gnre;
+        if (!c->itunes_metadata) {
+            int64_t pos = avio_tell(pb);
+            int ret = mov_read_3gp_udta_tag(c, pb, atom);
+            if (ret != AVERROR_INVALIDDATA)
+                return ret;
+            avio_seek(pb, pos, SEEK_SET);
+        }
+        break;
     case MKTAG( 'h','d','v','d'): key = "hd_video";
         parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 'H','M','M','T'):
         return mov_metadata_hmmt(c, pb, atom.size);
+    case MKTAG( 'i','t','n','u'): key = "itunes_u";
+        parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 'k','e','y','w'): key = "keywords";  break;
     case MKTAG( 'l','d','e','s'): key = "synopsis";  break;
     case MKTAG( 'l','o','c','i'):
         return mov_metadata_loci(c, pb, atom.size);
     case MKTAG( 'm','a','n','u'): key = "make"; break;
     case MKTAG( 'm','o','d','l'): key = "model"; break;
-    case MKTAG( 'n','a','m','e'): key = "name"; break;
+    case MKTAG( 'n','a','m','e'): key = "title"; raw = 1; break;
     case MKTAG( 'p','c','s','t'): key = "podcast";
         parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 'p','g','a','p'): key = "gapless_playback";
         parse = mov_metadata_int8_no_padding; break;
+    case MKTAG( 'p','l','I','D'): key = "playlist_id";
+        parse = mov_metadata_int64_no_padding; break;
     case MKTAG( 'p','u','r','d'): key = "purchase_date"; break;
     case MKTAG( 'r','t','n','g'): key = "rating";
         parse = mov_metadata_int8_no_padding; break;
+    case MKTAG( 's','f','I','D'): key = "itunes_country";
+        parse = mov_metadata_int32_no_padding; break;
+    case MKTAG( 's','d','e','s'): key = "series_description"; break;
+    case MKTAG( 's','h','w','m'): key = "show_work_and_movement";
+        parse = mov_metadata_int8_no_padding; break;
     case MKTAG( 's','o','a','a'): key = "sort_album_artist"; break;
     case MKTAG( 's','o','a','l'): key = "sort_album";   break;
     case MKTAG( 's','o','a','r'): key = "sort_artist";  break;
@@ -409,6 +539,8 @@
     case MKTAG( 's','o','s','n'): key = "sort_show";    break;
     case MKTAG( 's','t','i','k'): key = "media_type";
         parse = mov_metadata_int8_no_padding; break;
+    case MKTAG( 't','m','p','o'): key = "tmpo";
+        parse = mov_metadata_int16_no_padding; break;
     case MKTAG( 't','r','k','n'): key = "track";
         parse = mov_metadata_track_or_disc_number; break;
     case MKTAG( 't','v','e','n'): key = "episode_id"; break;
@@ -418,17 +550,23 @@
     case MKTAG( 't','v','s','h'): key = "show";      break;
     case MKTAG( 't','v','s','n'): key = "season_number";
         parse = mov_metadata_int8_bypass_padding; break;
+    case MKTAG( 'x','i','d',' '): key = "xid";       break;
     case MKTAG(0xa9,'A','R','T'): key = "artist";    break;
     case MKTAG(0xa9,'P','R','D'): key = "producer";  break;
     case MKTAG(0xa9,'a','l','b'): key = "album";     break;
-    case MKTAG(0xa9,'a','u','t'): key = "artist";    break;
+    case MKTAG(0xa9,'a','r','d'): key = "art_director"; break;
+    case MKTAG(0xa9,'a','r','g'): key = "arranger";  break;
+    case MKTAG(0xa9,'a','u','t'): key = "author";    break;
+    case MKTAG(0xa9,'c','a','k'): key = "acknowledgement"; break;
     case MKTAG(0xa9,'c','h','p'): key = "chapter";   break;
     case MKTAG(0xa9,'c','m','t'): key = "comment";   break;
     case MKTAG(0xa9,'c','o','m'): key = "composer";  break;
+    case MKTAG(0xa9,'c','o','n'): key = "conductor"; break;
     case MKTAG(0xa9,'c','p','y'): key = "copyright"; break;
     case MKTAG(0xa9,'d','a','y'): key = "date";      break;
     case MKTAG(0xa9,'d','i','r'): key = "director";  break;
     case MKTAG(0xa9,'d','i','s'): key = "disclaimer"; break;
+    case MKTAG(0xa9,'d','e','s'): key = "song_description"; break;
     case MKTAG(0xa9,'e','d','1'): key = "edit_date"; break;
     case MKTAG(0xa9,'e','n','c'): key = "encoder";   break;
     case MKTAG(0xa9,'f','m','t'): key = "original_format"; break;
@@ -436,23 +574,45 @@
     case MKTAG(0xa9,'g','r','p'): key = "grouping";  break;
     case MKTAG(0xa9,'h','s','t'): key = "host_computer"; break;
     case MKTAG(0xa9,'i','n','f'): key = "comment";   break;
+    case MKTAG(0xa9,'l','n','t'): key = "linear_notes"; break;
     case MKTAG(0xa9,'l','y','r'): key = "lyrics";    break;
     case MKTAG(0xa9,'m','a','k'): key = "make";      break;
     case MKTAG(0xa9,'m','o','d'): key = "model";     break;
+    case MKTAG(0xa9,'m','v','n'): key = "movement_name"; break;
+    case MKTAG(0xa9,'m','v','i'): key = "movement_number";
+        parse = mov_metadata_int16_no_padding; break;
+    case MKTAG(0xa9,'m','v','c'): key = "movement_count";
+        parse = mov_metadata_int16_no_padding; break;
     case MKTAG(0xa9,'n','a','m'): key = "title";     break;
     case MKTAG(0xa9,'o','p','e'): key = "original_artist"; break;
+    case MKTAG(0xa9,'p','h','g'): key = "phonogram_rights"; break;
     case MKTAG(0xa9,'p','r','d'): key = "producer";  break;
     case MKTAG(0xa9,'p','r','f'): key = "performers"; break;
+    case MKTAG(0xa9,'p','u','b'): key = "publisher"; break;
     case MKTAG(0xa9,'r','e','q'): key = "playback_requirements"; break;
+    case MKTAG(0xa9,'s','n','e'): key = "sound_engineer"; break;
+    case MKTAG(0xa9,'s','o','l'): key = "soloist";   break;
     case MKTAG(0xa9,'s','r','c'): key = "original_source"; break;
     case MKTAG(0xa9,'s','t','3'): key = "subtitle";  break;
     case MKTAG(0xa9,'s','w','r'): key = "encoder";   break;
+    case MKTAG(0xa9,'t','h','x'): key = "thanks";    break;
     case MKTAG(0xa9,'t','o','o'): key = "encoder";   break;
     case MKTAG(0xa9,'t','r','k'): key = "track";     break;
     case MKTAG(0xa9,'u','r','l'): key = "URL";       break;
+    case MKTAG(0xa9,'w','r','k'): key = "work_name"; break;
     case MKTAG(0xa9,'w','r','n'): key = "warning";   break;
     case MKTAG(0xa9,'w','r','t'): key = "composer";  break;
+    case MKTAG(0xa9,'x','p','d'): key = "executive_producer"; break;
     case MKTAG(0xa9,'x','y','z'): key = "location";  break;
+    case MKTAG( 'a','l','b','m'):
+    case MKTAG( 'a','u','t','h'):
+    case MKTAG( 'd','s','c','p'):
+    case MKTAG( 'p','e','r','f'):
+    case MKTAG( 't','i','t','l'):
+    case MKTAG( 'y','r','r','c'):
+        if (!c->itunes_metadata) {
+            return mov_read_3gp_udta_tag(c, pb, atom);
+        }
     }
 retry:
     if (c->itunes_metadata && atom.size > 8) {
@@ -9739,6 +9899,22 @@
     return ret;
 }
 
+static int mov_read_fall(MOVContext *c, AVIOContext *pb, MOVAtom atom)
+{
+    AVStream *st;
+    MOVStreamContext *sc;
+
+    if (c->fc->nb_streams < 1)
+        return 0;
+    st = c->fc->streams[c->fc->nb_streams-1];
+    sc = st->priv_data;
+
+    sc->fallback = avio_rb32(pb);
+    sc->has_fallback = 1;
+
+    return 0;
+}
+
 static const MOVParseTableEntry mov_default_parse_table[] = {
 { MKTAG('A','C','L','R'), mov_read_aclr },
 { MKTAG('A','P','R','G'), mov_read_avid },
@@ -9842,6 +10018,7 @@
 { MKTAG('v','p','c','C'), mov_read_vpcc },
 { MKTAG('m','d','c','v'), mov_read_mdcv },
 { MKTAG('c','l','l','i'), mov_read_clli },
+{ MKTAG('f','a','l','l'), mov_read_fall },
 { MKTAG('d','v','c','C'), mov_read_dvcc_dvvc },
 { MKTAG('d','v','v','C'), mov_read_dvcc_dvvc },
 { MKTAG('d','v','w','C'), mov_read_dvcc_dvvc },
@@ -11432,6 +11609,23 @@
             err = ff_replaygain_export(st, s->metadata);
             if (err < 0)
                 return err;
+            if (sc->has_fallback) {
+                for (j = 0; j < s->nb_streams; j++) {
+                    if (s->streams[j]->id == sc->fallback) {
+                        AVPacketSideData *sd;
+                        int *fallback;
+                        sd = av_packet_side_data_new(&st->codecpar->coded_side_data,
+                                    &st->codecpar->nb_coded_side_data,
+                                    AV_PKT_DATA_FALLBACK_TRACK,
+                                    sizeof(int), 0);
+                        if (!sd)
+                            return AVERROR(ENOMEM);
+                        fallback = (int*)sd->data;
+                        *fallback = j;
+                        break;
+                    }
+                }
+            }
             break;
         case AVMEDIA_TYPE_VIDEO:
             if (sc->display_matrix) {
diff -Naur ffmpeg-9.0.1.old/libavformat/movenc.c ffmpeg-9.0.1/libavformat/movenc.c
--- ffmpeg-9.0.1.old/libavformat/movenc.c	2026-09-18 07:59:38.477495883 +0200
+++ ffmpeg-9.0.1/libavformat/movenc.c	2026-09-18 07:59:39.660506877 +0200
@@ -28,6 +28,7 @@
 
 #include "movenc.h"
 #include "avformat.h"
+#include "avlanguage.h"
 #include "avio_internal.h"
 #include "dovi_isom.h"
 #include "riff.h"
@@ -4295,6 +4296,41 @@
     return len + 24;
 }
 
+static uint16_t language_code(const char *str)
+{
+    return (((str[0] - 0x60) & 0x1F) << 10) +
+           (((str[1] - 0x60) & 0x1F) <<  5) +
+           (( str[2] - 0x60) & 0x1F);
+}
+static int mov_write_3gp_udta_tag(AVIOContext *pb, AVDictionary *metadata,
+                                  const char *tag, const char *str)
+{
+    int64_t pos = avio_tell(pb);
+    AVDictionaryEntry *t = av_dict_get(metadata, str, NULL, 0);
+    if (!t || !utf8len(t->value))
+        return 0;
+    avio_wb32(pb, 0);   /* size */
+    ffio_wfourcc(pb, tag); /* type */
+    avio_wb32(pb, 0);   /* version + flags */
+    if (!strcmp(tag, "yrrc"))
+        avio_wb16(pb, atoi(t->value));
+    else {
+        int lang = 0, len;
+        len = strlen(t->key);
+        if (t->key[len - 4] == '-') {
+            lang = ff_mov_iso639_to_lang(&t->key[len - 3], 1);
+        }
+        if (!lang)
+            lang = ff_mov_iso639_to_lang("und", 1);
+        avio_wb16(pb, lang); /* language */
+        avio_write(pb, t->value, strlen(t->value) + 1); /* UTF8 string value */
+        if (!strcmp(tag, "albm") &&
+            (t = av_dict_get(metadata, "track", NULL, 0)))
+            avio_w8(pb, atoi(t->value));
+    }
+    return update_size(pb, pos);
+}
+
 static int mov_write_track_metadata(AVIOContext *pb, AVStream *st,
                                     const char *tag, const char *str)
 {
@@ -4369,8 +4405,23 @@
     if (ret < 0)
         return ret;
 
-    if (mov->mode & (MODE_MP4|MODE_MOV))
+    if (mov->mode & (MODE_MP4|MODE_MOV)) {
+        AVDictionaryEntry *t = NULL;
+        int und = 0;
+
         mov_write_track_metadata(pb_buf, st, "name", "title");
+        while ((t = av_dict_get(st->metadata, "title-", t, AV_DICT_IGNORE_SUFFIX))) {
+            int len = strlen(t->key);
+            if (len == 10 &&
+                ff_convert_lang_to(&t->key[len - 3], AV_LANG_ISO639_2_BIBL)) {
+                mov_write_3gp_udta_tag(pb_buf, st->metadata, "titl", t->key);
+                if (!strcmp("und", &t->key[len - 3]))
+                    und = 1;
+            }
+        }
+        if (!und)
+            mov_write_3gp_udta_tag(pb_buf, st->metadata, "titl", "title");
+    }
 
     if (mov->mode & MODE_MP4) {
         if ((ret = mov_write_track_kinds(pb_buf, st)) < 0)
@@ -4790,15 +4841,19 @@
                                    int len)
 {
     AVDictionaryEntry *t = NULL;
-    uint8_t num;
+    uint64_t num;
     int size = 24 + len;
 
-    if (len != 1 && len != 4)
+    if (len != 1 && len != 4 &&
+        len != 2 && len != 8)
         return -1;
 
     if (!(t = av_dict_get(s->metadata, tag, NULL, 0)))
         return 0;
-    num = atoi(t->value);
+    if (len <= 4)
+        num = atoi(t->value);
+    else
+        num = atol(t->value);
 
     avio_wb32(pb, size);
     ffio_wfourcc(pb, name);
@@ -4806,7 +4861,9 @@
     ffio_wfourcc(pb, "data");
     avio_wb32(pb, 0x15);
     avio_wb32(pb, 0);
-    if (len==4) avio_wb32(pb, num);
+    if (len==8) avio_wb64(pb, num);
+    else if (len==4) avio_wb32(pb, num);
+    else if (len==2) avio_wb16(pb, num);
     else        avio_w8 (pb, num);
 
     return size;
@@ -4862,6 +4919,8 @@
     mov_write_string_metadata(s, pb, "\251lyr", "lyrics"   , 1);
     mov_write_string_metadata(s, pb, "desc",    "description",1);
     mov_write_string_metadata(s, pb, "ldes",    "synopsis" , 1);
+    mov_write_string_metadata(s, pb, "sdes",    "series_description", 1);
+    mov_write_string_metadata(s, pb, "rtng",    "rating",    1);
     mov_write_string_metadata(s, pb, "tvsh",    "show"     , 1);
     mov_write_string_metadata(s, pb, "tven",    "episode_id",1);
     mov_write_string_metadata(s, pb, "tvnn",    "network"  , 1);
@@ -4872,10 +4931,67 @@
     mov_write_int8_metadata  (s, pb, "hdvd",    "hd_video",  1);
     mov_write_int8_metadata  (s, pb, "pgap",    "gapless_playback",1);
     mov_write_int8_metadata  (s, pb, "cpil",    "compilation", 1);
+
+    mov_write_string_metadata(s, pb, "\251st3", "subtitle"        , 1);
+    mov_write_string_metadata(s, pb, "\251des", "song_description", 1);
+    mov_write_string_metadata(s, pb, "\251dir", "director"        , 1);
+    mov_write_string_metadata(s, pb, "\251ard", "art_director"    , 1);
+    mov_write_string_metadata(s, pb, "\251arg", "arranger"        , 1);
+    mov_write_string_metadata(s, pb, "\251aut", "author"          , 1);
+    mov_write_string_metadata(s, pb, "\251cak", "acknowledgement" , 1);
+    mov_write_string_metadata(s, pb, "\251con", "conductor"       , 1);
+
+    mov_write_string_metadata(s, pb, "\251wrk", "work_name"     , 1);
+    mov_write_string_metadata(s, pb, "\251mvn", "movement_name" , 1);
+    mov_write_int8_metadata  (s, pb, "\251mvi", "movement_number", 2);
+    mov_write_int8_metadata  (s, pb, "\251mvc", "movement_count" , 2);
+    mov_write_int8_metadata  (s, pb, "shwm",    "show_work_and_movement", 1);
+
+    mov_write_string_metadata(s, pb, "\251lnt", "linear_notes"      , 1);
+    mov_write_string_metadata(s, pb, "\251mak", "make"              , 1); // Record company
+    mov_write_string_metadata(s, pb, "\251ope", "original_artist"   , 1);
+    mov_write_string_metadata(s, pb, "\251phg", "phonogram_rights"  , 1);
+    mov_write_string_metadata(s, pb, "\251prd", "producer"          , 1);
+    mov_write_string_metadata(s, pb, "\251prf", "performers"        , 1);
+    mov_write_string_metadata(s, pb, "\251pub", "publisher"         , 1);
+    mov_write_string_metadata(s, pb, "\251sne", "sound_engineer"    , 1);
+    mov_write_string_metadata(s, pb, "\251sol", "soloist"           , 1);
+    mov_write_string_metadata(s, pb, "\251src", "original_source"   , 1); // Credits
+    mov_write_string_metadata(s, pb, "\251thx", "thanks"            , 1);
+    mov_write_string_metadata(s, pb, "\251url", "URL"               , 1); // Online extras
+    mov_write_string_metadata(s, pb, "\251xpd", "executive_producer", 1);
+
+    mov_write_string_metadata(s, pb, "sonm", "sort_name"        , 1);
+    mov_write_string_metadata(s, pb, "soar", "sort_artist"      , 1);
+    mov_write_string_metadata(s, pb, "soaa", "sort_album_artist", 1);
+    mov_write_string_metadata(s, pb, "soal", "sort_album"       , 1);
+    mov_write_string_metadata(s, pb, "soco", "sort_composer"    , 1);
+    mov_write_string_metadata(s, pb, "sosn", "sort_show"        , 1);
+
+    mov_write_string_metadata(s, pb, "\251enc", "encoder"       , 1); // Encoded by
+    mov_write_string_metadata(s, pb, "purd",    "purchase_date" , 1);
+
+    mov_write_int8_metadata  (s, pb, "itnu",    "itunes_u"      , 1);
+    mov_write_int8_metadata  (s, pb, "pcst",    "podcast"       , 1);
+    mov_write_string_metadata(s, pb, "catg",    "category"      , 1);
+
+    mov_write_string_metadata(s, pb, "apID",    "account_id"    , 1);
+    mov_write_int8_metadata  (s, pb, "akID",    "account_type"  , 1);
+    mov_write_int8_metadata  (s, pb, "sfID",    "itunes_country", 4);
+    mov_write_int8_metadata  (s, pb, "cnID",    "content_id"    , 4);
+    mov_write_int8_metadata  (s, pb, "atID",    "artist_id"     , 4);
+    mov_write_int8_metadata  (s, pb, "plID",    "playlist_id"   , 8);
+    mov_write_int8_metadata  (s, pb, "geID",    "genre_id"      , 4);
+    mov_write_int8_metadata  (s, pb, "cmID",    "composer_id"   , 4);
+    mov_write_string_metadata(s, pb, "xid ",    "xid"           , 1);
+
     mov_write_covr(pb, s);
     mov_write_trkn_tag(pb, mov, s, 0); // track number
     mov_write_trkn_tag(pb, mov, s, 1); // disc number
     mov_write_tmpo_tag(pb, s);
+
+    mov_write_custom_metadata(s, pb, "com.apple.iTunes", "iTunEXTC", "iTunEXTC");
+    mov_write_custom_metadata(s, pb, "com.apple.iTunes", "iTunMOVI", "iTunMOVI");
     mov_write_custom_metadata(s, pb, "com.apple.iTunes", "DISCSUBTITLE", "disc_subtitle");
     return update_size(pb, pos);
 }
@@ -5008,35 +5124,6 @@
     return 0;
 }
 
-static uint16_t language_code(const char *str)
-{
-    return (((str[0] - 0x60) & 0x1F) << 10) +
-           (((str[1] - 0x60) & 0x1F) <<  5) +
-           (( str[2] - 0x60) & 0x1F);
-}
-
-static int mov_write_3gp_udta_tag(AVIOContext *pb, AVFormatContext *s,
-                                  const char *tag, const char *str)
-{
-    int64_t pos = avio_tell(pb);
-    AVDictionaryEntry *t = av_dict_get(s->metadata, str, NULL, 0);
-    if (!t || !utf8len(t->value))
-        return 0;
-    avio_wb32(pb, 0);   /* size */
-    ffio_wfourcc(pb, tag); /* type */
-    avio_wb32(pb, 0);   /* version + flags */
-    if (!strcmp(tag, "yrrc"))
-        avio_wb16(pb, atoi(t->value));
-    else {
-        avio_wb16(pb, language_code("eng")); /* language */
-        avio_write(pb, t->value, strlen(t->value) + 1); /* UTF8 string value */
-        if (!strcmp(tag, "albm") &&
-            (t = av_dict_get(s->metadata, "track", NULL, 0)))
-            avio_w8(pb, atoi(t->value));
-    }
-    return update_size(pb, pos);
-}
-
 static int mov_write_chpl_tag(AVIOContext *pb, AVFormatContext *s)
 {
     int64_t pos = avio_tell(pb);
@@ -5075,14 +5162,14 @@
         return ret;
 
     if (mov->mode & MODE_3GP) {
-        mov_write_3gp_udta_tag(pb_buf, s, "perf", "artist");
-        mov_write_3gp_udta_tag(pb_buf, s, "titl", "title");
-        mov_write_3gp_udta_tag(pb_buf, s, "auth", "author");
-        mov_write_3gp_udta_tag(pb_buf, s, "gnre", "genre");
-        mov_write_3gp_udta_tag(pb_buf, s, "dscp", "comment");
-        mov_write_3gp_udta_tag(pb_buf, s, "albm", "album");
-        mov_write_3gp_udta_tag(pb_buf, s, "cprt", "copyright");
-        mov_write_3gp_udta_tag(pb_buf, s, "yrrc", "date");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "perf", "artist");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "titl", "title");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "auth", "author");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "gnre", "genre");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "dscp", "comment");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "albm", "album");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "cprt", "copyright");
+        mov_write_3gp_udta_tag(pb_buf, s->metadata, "yrrc", "date");
         mov_write_loci_tag(s, pb_buf);
     } else if (mov->mode == MODE_MOV && !(mov->flags & FF_MOV_FLAG_USE_MDTA)) { // the title field breaks gtkpod with mp4 and my suspicion is that stuff is not valid in mp4
         mov_write_string_metadata(s, pb_buf, "\251ART", "artist",      0);
@@ -7757,6 +7844,7 @@
     track->tag = MKTAG('t','e','x','t');
     track->timescale = mov->movie_timescale;
     track->par = avcodec_parameters_alloc();
+    track->language = mov->tracks[0].language;
     if (!track->par)
         return AVERROR(ENOMEM);
     track->par->codec_type = AVMEDIA_TYPE_SUBTITLE;
diff -Naur ffmpeg-9.0.1.old/libavutil/frame.h ffmpeg-9.0.1/libavutil/frame.h
--- ffmpeg-9.0.1.old/libavutil/frame.h	2026-09-18 07:59:38.427495418 +0200
+++ ffmpeg-9.0.1/libavutil/frame.h	2026-09-18 07:59:39.655348723 +0200
@@ -201,6 +201,12 @@
     AV_FRAME_DATA_DOVI_RPU_BUFFER,
 
     /**
+     * Dolby Vision RPU ITU T35 raw data, suitable for passing to SVT-AV1
+     * or other libraries. Array of uint8_t.
+     */
+    AV_FRAME_DATA_DOVI_RPU_BUFFER_T35,
+
+    /**
      * Parsed Dolby Vision metadata, suitable for passing to a software
      * implementation. The payload is the AVDOVIMetadata struct defined in
      * libavutil/dovi_meta.h.
diff -Naur ffmpeg-9.0.1.old/libavutil/hwcontext_d3d11va.c ffmpeg-9.0.1/libavutil/hwcontext_d3d11va.c
--- ffmpeg-9.0.1.old/libavutil/hwcontext_d3d11va.c	2026-09-18 07:59:38.438284140 +0200
+++ ffmpeg-9.0.1/libavutil/hwcontext_d3d11va.c	2026-09-18 07:59:39.666381733 +0200
@@ -85,6 +85,9 @@
     int nb_surfaces;
     int nb_surfaces_used;
 
+    int max_retries;
+    int retries;
+
     DXGI_FORMAT format;
 
     ID3D11Texture2D *staging_texture;
@@ -261,7 +264,9 @@
     ID3D11Texture2D_GetDesc(hwctx->texture, &texDesc);
 
     if (s->nb_surfaces_used >= texDesc.ArraySize) {
-        av_log(ctx, AV_LOG_ERROR, "Static surface pool size exceeded.\n");
+        if (s->retries >= s->max_retries) {
+            av_log(ctx, AV_LOG_ERROR, "Static surface pool size exceeded.\n");
+        }
         return NULL;
     }
 
@@ -347,20 +352,34 @@
 static int d3d11va_get_buffer(AVHWFramesContext *ctx, AVFrame *frame)
 {
     AVD3D11FrameDescriptor *desc;
+    D3D11VAFramesContext       *s = ctx->hwctx;
+    s->retries = 0;
+    s->max_retries = 50;
 
-    frame->buf[0] = av_buffer_pool_get(ctx->pool);
-    if (!frame->buf[0])
-        return AVERROR(ENOMEM);
-
-    desc = (AVD3D11FrameDescriptor *)frame->buf[0]->data;
-
-    frame->data[0] = (uint8_t *)desc->texture;
-    frame->data[1] = (uint8_t *)desc->index;
-    frame->format  = AV_PIX_FMT_D3D11;
-    frame->width   = ctx->width;
-    frame->height  = ctx->height;
+    /**
+     * Loop until a buffer becomes available from the pool.
+     * In a full hardware pipeline, all buffers may be temporarily in use by
+     * other modules (encoder/filter/decoder). Rather than immediately failing
+     * with ENOMEM, we wait for a buffer to be released back to the pool, which
+     * maintains pipeline flow and prevents unnecessary allocation failures
+     * during normal operation.
+     */
+    while (s->retries < s->max_retries) {
+        frame->buf[0] = av_buffer_pool_get(ctx->pool);
+        if (frame->buf[0]) {
+            desc = (AVD3D11FrameDescriptor *)frame->buf[0]->data;
+            frame->data[0] = (uint8_t *)desc->texture;
+            frame->data[1] = (uint8_t *)desc->index;
+            frame->format  = AV_PIX_FMT_D3D11;
+            frame->width   = ctx->width;
+            frame->height  = ctx->height;
+            return 0;
+        }
 
-    return 0;
+        av_usleep(1000);
+    }
+ 
+    return AVERROR(ENOMEM);
 }
 
 static int d3d11va_transfer_get_formats(AVHWFramesContext *ctx,
diff -Naur ffmpeg-9.0.1.old/libavutil/hwcontext_videotoolbox.c ffmpeg-9.0.1/libavutil/hwcontext_videotoolbox.c
--- ffmpeg-9.0.1.old/libavutil/hwcontext_videotoolbox.c	2026-09-18 07:59:38.438495521 +0200
+++ ffmpeg-9.0.1/libavutil/hwcontext_videotoolbox.c	2026-09-18 07:59:39.664359030 +0200
@@ -194,7 +194,8 @@
     VTFramesContext *fctx = ctx->hwctx;
     AVVTFramesContext *hw_ctx = &fctx->p;
     CVReturn err;
-    CFNumberRef w, h, pixfmt;
+    CFNumberRef w, h, extend_w_num, extend_h_num, pixfmt;
+    int extend_w, extend_h;
     uint32_t cv_pixfmt;
     CFMutableDictionaryRef attributes, iosurface_properties;
 
@@ -227,6 +228,15 @@
     CFRelease(w);
     CFRelease(h);
 
+    extend_w = FFALIGN(ctx->width,  16) - ctx->width;
+    extend_h = FFALIGN(ctx->height, 16) - ctx->height;
+    extend_w_num = CFNumberCreate(NULL, kCFNumberSInt32Type, &extend_w);
+    extend_h_num = CFNumberCreate(NULL, kCFNumberSInt32Type, &extend_h);
+    CFDictionarySetValue(attributes, kCVPixelBufferExtendedPixelsRightKey, extend_w_num);
+    CFDictionarySetValue(attributes, kCVPixelBufferExtendedPixelsBottomKey, extend_h_num);
+    CFRelease(extend_w_num);
+    CFRelease(extend_h_num);
+
     err = CVPixelBufferPoolCreate(
         NULL,
         NULL,
diff -Naur ffmpeg-9.0.1.old/libavutil/side_data.c ffmpeg-9.0.1/libavutil/side_data.c
--- ffmpeg-9.0.1.old/libavutil/side_data.c	2026-09-18 07:59:38.426495409 +0200
+++ ffmpeg-9.0.1/libavutil/side_data.c	2026-09-18 07:59:39.655446787 +0200
@@ -42,6 +42,7 @@
     [AV_FRAME_DATA_FILM_GRAIN_PARAMS]           = { "Film grain parameters" },
     [AV_FRAME_DATA_DETECTION_BBOXES]            = { "Bounding boxes for object detection and classification", AV_SIDE_DATA_PROP_SIZE_DEPENDENT },
     [AV_FRAME_DATA_DOVI_RPU_BUFFER]             = { "Dolby Vision RPU Data",                        AV_SIDE_DATA_PROP_COLOR_DEPENDENT },
+    [AV_FRAME_DATA_DOVI_RPU_BUFFER_T35]         = { "Dolby Vision RPU ITU T35 Data",                AV_SIDE_DATA_PROP_COLOR_DEPENDENT },
     [AV_FRAME_DATA_DOVI_METADATA]               = { "Dolby Vision Metadata",                        AV_SIDE_DATA_PROP_COLOR_DEPENDENT },
     [AV_FRAME_DATA_LCEVC]                       = { "LCEVC NAL data",                               AV_SIDE_DATA_PROP_SIZE_DEPENDENT },
     [AV_FRAME_DATA_VIEW_ID]                     = { "View ID" },
diff -Naur ffmpeg-9.0.1.old/libswresample/rematrix.c ffmpeg-9.0.1/libswresample/rematrix.c
--- ffmpeg-9.0.1.old/libswresample/rematrix.c	2026-09-18 07:59:38.450455727 +0200
+++ ffmpeg-9.0.1/libswresample/rematrix.c	2026-09-18 07:59:39.667576626 +0200
@@ -110,7 +110,7 @@
         return 0;
     if(ch_layout->order == AV_CHANNEL_ORDER_CUSTOM)
         for (int i = 0; i < ch_layout->nb_channels; i++) {
-            if (ch_layout->u.map[i].id >= 64)
+            if (ch_layout->u.map[i].id >= 64 && ch_layout->u.map[i].id != AV_CHAN_UNUSED)
                 return 0;
         }
     else if (ch_layout->order != AV_CHANNEL_ORDER_NATIVE)
