diff -Naur ffmpeg-7.1.4.old/configure ffmpeg-7.1.4/configure
--- ffmpeg-7.1.4.old/configure	2026-05-05 08:18:38.444481156 +0200
+++ ffmpeg-7.1.4/configure	2026-05-05 08:18:38.447424332 +0200
@@ -4710,6 +4710,9 @@
     if $nvcc $nvccflags_default 2>&1 | grep -qi unsupported; then
         nvccflags_default="-gencode arch=compute_60,code=sm_60 -O2"
     fi
+    if $nvcc $nvccflags_default 2>&1 | grep -qi unsupported; then
+        nvccflags_default="-gencode arch=compute_75,code=sm_75 -O2"
+    fi
 fi
 
 set_default arch cc cxx doxygen pkg_config ranlib strip sysinclude \
