diff -Naur ffmpeg-6.0.1.old/configure ffmpeg-6.0.1/configure
--- ffmpeg-6.0.1.old/configure	2023-11-21 13:26:44.841753682 +0100
+++ ffmpeg-6.0.1/configure	2023-11-21 13:26:45.114758335 +0100
@@ -4453,10 +4453,10 @@
 
 if enabled cuda_nvcc; then
     nvcc_default="nvcc"
-    nvccflags_default="-gencode arch=compute_30,code=sm_30 -O2"
+    nvccflags_default="-allow-unsupported-compiler -gencode arch=compute_52,code=sm_52 -O2"
 else
     nvcc_default="clang"
-    nvccflags_default="--cuda-gpu-arch=sm_30 -O2"
+    nvccflags_default="--cuda-gpu-arch=sm_52 -O2"
     NVCC_C=""
 fi
 
